### ISSUE TITLE
Matrices subtraction dimension error 

### ISSUE THREAD

**Issue by JasonShin** [josdejong/mathjs#1165]
I am getting a dimension error while using math.subtract()

```
const X = [[1, 2], [3, 4], [5, 6]]
const A = math.transpose(X);
// A = [ [ 1, 3, 5 ], [ 2, 4, 6 ] ] 
const M = math.mean(A, 1);
// M = [ 3, 4 ]
const C = math.subtract(X, M);
```

However, it throws

```
....mathjs/lib/function/arithmetic/subtract.js:172
    throw new DimensionError(xsize.length, ysize.length);

```

I believe `math.subtract(X, M)` should return

```
[[-2., -2.],
[ 0.,  0.],
[ 2.,  2.]]
```

same goes for `math.sum` `math.multiply`

**Comment by JasonShin** [comment]
In the meantime, I will write a extension method to deal with it

**Comment by josdejong** [comment]
Yes, that can be a useful extension. Right now `math.subtract`, `math.add`, etc simply do element-wise subtraction/addition. Allowing to subtract a vector row-wise from a matrix is something you can do in Matlab and Octave. 

I will change the title of this topic a bit.

Anyone interested in implementing this feature?

**Comment by JasonShin** [comment]
Once I finish the `getMatrixDataType` pull request, I am happy to implement it =)

**Comment by ChristopherChudzicki** [comment]
This could definitely be useful in some cases. But  I suggest that there should be a way to turn it off.

Example: I mainly use MathJS for math webapps to be used by high-school and undergraduate students. I generally want students to think that adding matrix + vector is not valid. Great, MathJS throws a useful error!

I could be one of a small minority, but I would guess that MathJS's current behavior is useful to other people.

(Admittedly, if people like me want this error, it's not too hard to do our own validation. But that makes upgrading harder.)

Incidentally, `numpy` calls this  [broadcasting](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.htm). (Although, Numpy's broadcasting is more general.)

**Comment by JasonShin** [comment]
Thanks for an input =)

**Comment by josdejong** [comment]
Thanks for sharing your thoughts @ChristopherChudzicki. Very good point to weight convenience against more explicit notations preventing you from accidental mistakes. We could also think about a solution where we have separate, explicit functions (like `subtractVector(matrix, vector)`) to do operations on matrices and vectors, or pass an extra option to these functions 
(like `subtract(matrix, vector, { broadcast: true })` 
or `subtract(matrix, vector, { rowWise: true })`). 

Just thinking aloud here :)

**Comment by ChristopherChudzicki** [comment]
That said, it occurs to me that allowing `matrix + row` or `matrix + vector` consistent with how MathJS handles `matrix + scalar`. What I mean is, in the `matrix+scalar` case MathJS already does not demand strict shape-matching.

**Comment by josdejong** [comment]
Yes that's true.

I'm just exploring different options, not sure yet what solution would best serve the audience of mathjs.

**Comment by chivington** [comment]
Any progress on this? Am getting into mathjs as a numpy replacement in the browser for implementing ML/AI stuffs. [Broadcasting](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html) is a verrry handy feature that makes a lot of learning algorithms a bit simpler to implement. Happy to pitch in.

**Comment by chivington** [comment]
Also might add the word "Broadcasting" to the title for the numpy folks looking for help on this. I'd imagine they'll become a larger and larger part of your target audience, if they aren't currently.

**Comment by josdejong** [comment]
@chivington thanks for your input. There is no progress on this feature so far. I've marked the issue with a label "help wanted", now.

Anyone interested in implementing this feature?

**Comment by joeGermany** [comment]
Hello!! My case is similar to @chivington. I am getting into machine learning on the web and I need an alternative for NumPy. The above feature (broadcasting) is a real help for easier manipulation of matrices. Is there any way I can implement it alone in the meantime until it is a mainstream MathJS function (if it will ever be). 😃

**Comment by josdejong** [comment]
> Is there any way I can implement it alone in the meantime until it is a mainstream MathJS function (if it will ever be). 

The solution would be to implement your own version of all operators/functions like subtract, add, etc, like Jason suggests in https://github.com/josdejong/mathjs/issues/1165#issuecomment-404493249.
